File: 2095-delete-the-middle-node-of-a-linked-list/2095-delete-the-middle-node-of-a-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def deleteMiddle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        #Approach2
        temp = head
        if head is None:
            return head
        if head.next is None:
            return None
        c = 0
        while temp is not None:
            c += 1
            temp = temp.next
        temp = head
        for i in range(c//2-1):
            temp = temp.next
        temp.next = temp.next.next
        return head
        #Approach 1
#         slow, fast = head, head
#         if head is None:
#             return head
#         if head.next is None:
#             return None
#         c = 0
#         while fast and fast.next:
#             c += 1
#             if c == 1:
#                 prev = head
#                 slow = slow.next
#                 fast = fast.next.next
#             else:
#                 prev = prev.next
#                 slow = slow.next
#                 fast = fast.next.next
#         prev.next = slow.next
        
#         return head

        

File: 2095-delete-the-middle-node-of-a-linked-list/2095-delete-the-middle-node-of-a-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* deleteMiddle(ListNode* head) {
        ListNode *slow = head;
        ListNode* fast = head;
        ListNode *prev = nullptr;
        if(head == nullptr){
            return nullptr;
        }
        if(head->next == nullptr){
            return nullptr;
        }
        
        int c = 0;
        while(fast != nullptr && fast -> next!= nullptr){
            c ++;
            if(c == 1){
                prev = head;
                slow = slow->next;
                fast = fast -> next -> next;
                
            }else{
                prev = prev -> next;
                slow = slow -> next;
                fast = fast -> next -> next;
            }
            
        }
        prev -> next = slow ->next;
        return head;
    }
};